components/ws-manager/pkg/manager/pod_controller.go

```diff
@@ -67,6 +68,21 @@ func (r *PodReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.R
 // SetupWithManager sets up the controller with the Manager.
 func (r *PodReconciler) SetupWithManager(mgr ctrl.Manager) error {
 	return ctrl.NewControllerManagedBy(mgr).
+		WithEventFilter(
+			predicate.Funcs{
+				UpdateFunc: func(e event.UpdateEvent) bool {
+					_, ok := e.ObjectNew.GetAnnotations()[workspaceIDAnnotation]
+					return ok
+				},
+				CreateFunc: func(e event.CreateEvent) bool {
+					_, ok := e.Object.GetAnnotations()[workspaceIDAnnotation]
+					return ok
+				},
+				DeleteFunc: func(e event.DeleteEvent) bool {
+					_, ok := e.Object.GetAnnotations()[workspaceIDAnnotation]
+					return ok
+				},
```

Comment:
We can remove [these lines](https://github.com/gitpod-io/gitpod/blob/7136c68be3b0409473d351a88dd68c41e1ae5be7/components/ws-manager/pkg/manager/pod_controller.go#L41-L44) now.

Reply:
Since we have to get `queue`  val, we can  't delete these lines
https://github.com/gitpod-io/gitpod/blob/7136c68be3b0409473d351a88dd68c41e1ae5be7/components/ws-manager/pkg/manager/pod_controller.go#L45

Reply:
Given that you added the event filter.
If the workspace ID annotation does not exist, this event would not proceed anymore.
https://github.com/kubernetes-sigs/controller-runtime/blob/3c4deba74e5cdf647e232c5c1bbffcb14e2fb74f/pkg/predicate/predicate.go#L57-L64

Reply:
`workpool.go`'s Add method requires a [queue](https://github.com/gitpod-io/gitpod/blob/e0bb385493db33224ab63fd3623f8afa62296f3e/components/ws-manager/pkg/manager/internal/workpool/workpool.go#L80), which is the instanceID of the workspace. If we leave the lines (which are cheap to keep), we'll never send empty queue to workpool. It's unlikely, but, a pod could be patched with json specification to have the `gitpod/id` annotation's value set to empty string.

@jenting 
suggestion (non-blocking): if its desired to remove these lines, let's do in a follow-up PR, so as to not block this from the next deployment.